pkg/controller/plan/adapter/vsphere/builder.go

```diff
@@ -1034,32 +1045,213 @@ func (r *Builder) LunPersistentVolumeClaims(vmRef ref.Ref) (pvcs []core.Persiste
 	return
 }
 
+// FIXME rgolan - the behaviour needs to be per disk hense this method is flawed. Needs a bigger change.
+// For now this method returns true, if there's a mapping (backend by copy-offload-mapping ConfigMap, that
+// maps StoragetClasses to Vsphere data stores
 func (r *Builder) SupportsVolumePopulators() bool {
-	return false
+	mapping, err := r.GetCopyOffloadMapping()
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			// no support for copy offload or not configured. The config map
+			// is a must for the feature to work
+			r.Log.Info("the config map to configure copy offloading is missing, hence the feature is unusable", CopyOffloadConfigMap)
+		}
+		return false
+	}
+
+	return len(mapping.StorageClasses) > 0
 }
 
+// PopulatorVolumes is needed for vSphereXcopyVolomePopulator
 func (r *Builder) PopulatorVolumes(vmRef ref.Ref, annotations map[string]string, secretName string) (pvcs []*core.PersistentVolumeClaim, err error) {
-	err = planbase.VolumePopulatorNotSupportedError
-	return
+	vm := &model.VM{}
+	err = r.Source.Inventory.Find(vm, vmRef)
+	if err != nil {
+		err = liberr.Wrap(err, "vm", vmRef.String())
+		return
+	}
+
+	copyOffloadMapping, err := r.GetCopyOffloadMapping()
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			// no support for copy offload or not configured. The config map
+			// is a must for the feature to work
+		} else {
+			return nil, err
+		}
+	}
+
+	r.Log.Info("copy-offload capable storage classes", "storageClasses", copyOffloadMapping)
+	dsMapIn := r.Context.Map.Storage.Spec.Map
+	for i := range dsMapIn {
+		mapped := &dsMapIn[i]
+		ref := mapped.Source
+		ds := &model.Datastore{}
+		fErr := r.Source.Inventory.Find(ds, ref)
+		if fErr != nil {
+			err = fErr
+			return
+		}
+
+		pvblock := core.PersistentVolumeBlock
+		for _, disk := range vm.Disks {
+			if disk.Datastore.ID == ds.ID {
+				storageClass := mapped.Destination.StorageClass
+				coldLocal, vErr := r.Context.Plan.VSphereColdLocal()
+				if vErr != nil {
+					err = vErr
+					return
+				}
+				r.Log.Info(fmt.Sprintf("getting storage mapping by storage class %q and datastore %v datastore name %s datastore", storageClass, disk.Datastore, disk.Datastore))
+				vsphereInstance, storageVendorProduct, storageVendorSecretRef := copyOffloadMapping.GetStorageMappingBy(storageClass, ds.Name)
+				if coldLocal && vsphereInstance != "" {
+					namespace := r.Plan.Spec.TargetNamespace
+					// pvs names needs to be less than 63, this leaves 53 chars
+					// for the plan and vm name (2 dashes and 8 chars uuid)
+					commonName := fmt.Sprintf("%s-%s-%s", r.Plan.Name, vm.Name, uuid.New().String()[:8])
+					labels := map[string]string{
+						"migration": string(r.Migration.UID),
+						// we need uniqness and a value which is less than 64 chars, hence using disk.key
+						"vmdkKey": fmt.Sprint(disk.Key),
```

Comment:
note: the disk.key is not that unique by itself as the IDs easily repeat for each VM. but with the `vmRef.ID` this combination works well